File: client/mutations/create_contestant.js
import Relay, { Mutation } from 'react-relay'

export default class CreateContestantMutation extends Mutation {
  getMutation() {
    return Relay.QL`mutation { create_contestant }`
  }

  getVariables() {
    return {
      pool_id: this.props.pool_id,
      first_name: this.props.first_name,
      last_name: this.props.last_name,
      residence: this.props.residence,
      description: this.props.description
    }
  }

  getFatQuery() {
    return Relay.QL`
      fragment on CreateContestantPayload {
        contestant_edge
        pool
      }
    `
  }

  getConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentName: 'pool',
      parentID: this.props.pool_id,
      connectionName: 'contestants',
      edgeName: 'contestant_edge',
      rangeBehaviors: {
        '': 'append'
      }
    }]
  }
}

File: app/graph/mutations/create_contestant.rb
module Mutations
  CreateContestant = GraphQL::Relay::Mutation.define do
    name "CreateContestant"

    input_field :pool_id, !types.ID
    input_field :first_name, !types.String
    input_field :last_name, types.String
    input_field :residence, types.String
    input_field :description, types.String

    return_field :contestant_edge, Queries::ContestantType.edge_type
    return_field :pool, Queries::PoolType

    resolve ->(inputs, _context) do
      pool = Queries::NodeIdentification.object_from_id(inputs[:pool_id], {})
      contestant = pool.contestants.create!(
        first_name: inputs[:first_name],
        last_name: inputs[:last_name],
        residence: inputs[:residence],
        description: inputs[:description]
      )

      connection = GraphQL::Relay::RelationConnection.new(pool.contestants, {})

      {
        contestant_edge: GraphQL::Relay::Edge.new(contestant, connection),
        pool: pool
      }
    end
  end
end
